Create FixedLengthFrameDecoder

FranzBusch opened the pull request:
Created the LengthDelimitedInboundHandler

### Motivation:

https://github.com/apple/swift-nio/issues/474

normanmaurer commented:
@swift-nio-bot test this please.

normanmaurer commented:
@Lukasa @weissi do we move these handlers to an extra module (as we will have soon a few of these) ? That said I think we can merge this one first and then move ourselves to keep things simple.

normanmaurer commented:
@swift-nio-bot test this please

weissi commented on Sources/NIO/FixedLengthFrameDecoder.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        guard let slice = buffer.readSlice(length: frameLength) else {
```
we should do something here when the handler is removed and `buffer` is non-empty. Options that I can currently think of:

- fire the remaining bytes through the pipeline by implementing `handlerRemoved` as `ctx.fireChannelRead(self.wrapInboundOut(self.buffer)); ctx.fireChannelReadComplete()`
- fire an error that we have some left-over

CC @normanmaurer / @Lukasa

Lukasa replied:
Why not split the difference and fire an error that contains the left-over bytes?

weissi replied:
@Lukasa yes, that's also what I think we should do. Maybe we should have a global 'left over stuff' error type with an associated ByteBuffer?

Lukasa replied:
Sounds good to me.

weissi commented on Tests/NIOTests/FixedLengthFrameDecoderTest+XCTest.swift:
```diff
@@ -0,0 +1,34 @@
+
+   static var allTests : [(String, (FixedLengthFrameDecoderTest) -> () throws -> Void)] {
+      return [
+                ("testDecodeIfLessBytesAreSend", testDecodeIfLessBytesAreSend),
```
super minor but: this should be `...FewerBytesAreSent` not `...LessBytesAreSend`

weissi commented on Tests/NIOTests/FixedLengthFrameDecoderTest+XCTest.swift:
```diff
@@ -0,0 +1,34 @@
+   static var allTests : [(String, (FixedLengthFrameDecoderTest) -> () throws -> Void)] {
+      return [
+                ("testDecodeIfLessBytesAreSend", testDecodeIfLessBytesAreSend),
+                ("testDecodeIfMoreBytesAreSend", testDecodeIfMoreBytesAreSend),
```
this should also be `Sent` not `Send`

weissi commented on Tests/NIOTests/FixedLengthFrameDecoderTest.swift:
```diff
@@ -0,0 +1,55 @@
+        let channel = EmbeddedChannel()
+
+        let frameLength = 8
+        _ = try channel.pipeline.add(handler: FixedLengthFrameDecoder(frameLength: frameLength)).wait()
```
remove the `_ =` that's not needed as the type should be `Void` so you don't need to ignore it.

weissi commented on Tests/NIOTests/FixedLengthFrameDecoderTest.swift:
```diff
@@ -0,0 +1,55 @@
+        let channel = EmbeddedChannel()
+
+        let frameLength = 8
+        _ = try channel.pipeline.add(handler: FixedLengthFrameDecoder(frameLength: frameLength)).wait()
```
same here, remove `_ = `

weissi commented on Tests/NIOTests/FixedLengthFrameDecoderTest.swift:
```diff
@@ -0,0 +1,55 @@
+        XCTAssertEqual("aaaaaaaa", outputBuffer?.readString(length: frameLength))
+
+        outputBuffer = channel.readInbound()
+        XCTAssertNil(outputBuffer?.readString(length: frameLength))
```
we should call `XCTAssertFalse(channel.finish())` here, same above

weissi commented:
@normanmaurer / @Lukasa yeah, we should make our mind up where we put these handlers. Extra module or maybe in the [`swift-nio-extras`](https://github.com/apple/swift-nio-extras) repo?

ianpartridge commented:
This looks like a useful addition but I'm a bit confused - you reference #474 but that describes decoding a length-delimited framing protocol, not a fixed length frame decoder - they are different things. Probably the title and description of this PR needs updating?

Lukasa commented:
Yeah, the same issue was interpreted in two different ways. Conveniently, both are useful!

FranzBusch commented:
@weissi @Lukasa I updated the PR with the latest code review comments and implemented the `handlerRemoved` method which throws a new error `removedWithLeftOverBytes` if anything is left in the `cumulationBuffer`

weissi commented on Sources/NIO/FixedLengthFrameDecoder.swift:
```diff
@@ -0,0 +1,65 @@
+/// A decoder that splits the received `ByteBuffer` by a fixed number
+/// of bytes. For example, if you received the following four fragmented packets:
+///
+/// +---+----+------+----+
```
super minor (feel free to ignore): add 4 extra spaces before that so it gets code formatting

weissi commented on Sources/NIO/FixedLengthFrameDecoder.swift:
```diff
@@ -0,0 +1,65 @@
+/// A `FixedLengthFrameDecoder` will decode them into the
+/// following three packets with the fixed length:
+///
+/// +-----+-----+-----+
```
super minor (feel free to ignore): add 4 extra spaces before that so it gets code formatting

weissi commented on Sources/NIO/ChannelPipeline.swift:
```diff
@@ -989,11 +989,13 @@ private extension CloseMode {
 
 /// `Error` that is used by the `ChannelPipeline` to inform the user of an error.
-public enum ChannelPipelineError: Error {
+public enum ChannelPipelineError: Error, Equatable {
```
unfortunately because of SemVer we cannot extend this `enum` with extra cases before 2.0.0 is released. Therefore please add a separate type for this error. It could literally just be

```public struct LeftOverBytesError: Error {
    public let leftOverBytes: ByteBuffer
}```

CC @normanmaurer / @Lukasa for input about the name

FranzBusch commented:
@weissi Okay, moved the error to a separate struct. Was that what you had in mind?

Lukasa commented on Sources/NIO/FixedLengthFrameDecoder.swift:
```diff
@@ -0,0 +1,65 @@
+///     +-----+-----+-----+
+///
+public final class FixedLengthFrameDecoder: ByteToMessageDecoder {
+
```
Nit: remove this whitespace line.

Lukasa commented on Sources/NIO/FixedLengthFrameDecoder.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        guard let buffer = cumulationBuffer, buffer.readableBytes > 0 else {
```
Phrasing this as `guard let` reads weirdly to me, because that's not the expected flow. It might be better to rephrase it as `if let`.

Lukasa commented:
I believe @weissi has further code review here.

weissi commented:
@Lukasa I think I'm good for now actually. Only question is if this should go into `NIO` itself or [`swift-nio-extras`](https://github.com/apple/swift-nio-extras). It feels more like a swift-nio-extras thing to me but your call.

Lukasa commented:
Yeah, let's move this over to `swift-nio-extras`. @FranzBusch do you mind?
